### ISSUE TITLE
Limited brightness range 0...100 does not allow very low dim values

### ISSUE THREAD

**Issue by Strunzdesign** [ioBroker/ioBroker.zigbee#600]
For each of my Philips Hue bulbs (LCT012 and others) the Hue bridge provides both a "level" state and a "brightness" state. The "brightness state" has a range from 0...254 but the "level state" is mapped to a range of 0...100, certainly to match to dimmers providing a dim value in percent.

The Zigbee adaptor bases communication to the bulb on the "brightness values" (0...254), but performs an internal calculation to only provide "level-like values" to the ioBroker object tree. This has a drawback, as the lowest possible "level value" 1% is mapped to a "brightness value" of 3. I use the bulb as night lights, where a "brightness value" of 1 provides less light than the Zigbee adaptor currently allows.

Unfortunately, as this logic was introduced (https://github.com/ioBroker/ioBroker.zigbee/commit/65fb4f82b92b2cbeeb62c6bc13e83e2efc38dd4a), the name of the state was kept at "brightness", instead of renaming it to the more appropriate name "level".

My proposal: I would like to implement the following feature and offer a pull request, if nobody disagrees:

- Rename the current "brightness" state to the more appropriate state name "level" [0...100]
- Provide a "brightness" state providing access to the "raw brightness" value [0...254]
- Add code to always react to changes of / populate the level state and the brightness state
- Problem: scripts expecting a "brightness" range from (0...100) need to be changed to use the "level" state
- This would align the behavior of the Zigbee adaptor to that of the Hue Bridge adaptor, which always offers both states per bulb (level + "bri" brightness)

Regards,
Florian

**Comment by allofmex** [comment]
Oha, I always thought, 0-100 brightness uses the full range, but you are right, 1% sets value 3.
This for sure should be changed, 1% must set the lowest possible value (1% = 1, 100%=255).

But lets **keep** the name brightness, a change would require anyone with any type of bulb to adjust the scripts. 

On the other hand: If brightness state still uses percent values but maps to full range 1-255 - Do we need a raw state at all? (we may miss a few raw levels in between, but as long as the low levels can be fully addressed (1%=1, 2%=2, lets say 1:1 until 20%), it will not make much difference on higher level if 80%=200, 81%=203.

(Sadly I don't see a difference if I set a GU10 Hue bulb to level 1 and the other to level 3 with DeveloperTab)

**Comment by Strunzdesign** [comment]
Hi allofmex,

you are right, changing the name is not a practical step, even if the better choice. Given that, your proposal to remap "level" 1 to "brightness" 1 should do the trick. The differences between raw brightness levels are nearly not visible.

Thus, a linear mapping from "Zigbee brightness" [1...100] to "bulb brightness" [1...254] is desired, with 0 is 0 for "off " in both worlds. A raw value of brightness "1" provided by a bulb shall also be converted to a "Zigbee brightness" of "1", which is currently rounded down to 0.

Seems fine, not perfect, but practical.

**Comment by Strunzdesign** [comment]
Who is going to implement this? I can do it and offer a pull request... or do you want to do that?

**Comment by allofmex** [comment]
Would be nice if you can do it. Have no time for it..

**Comment by Strunzdesign** [comment]
Fixed due to merged pull request. Closing.

**Comment by Strunzdesign** [comment]
There is still an issue with a brightness level 1 set from the admin interface. It is converted correctly to a bulb brightness level of 1 as expected, but that value seems to switch my bulbs off instead of keeping it on at a very low level.

No idea, but it seems that a brightness level of "1" may be a reserved value, or this value is handled incorrectly somewhere down the road, maybe somehere inside the herdsman lib.

Can a developer please test this, setting a brightness level of "1" while watching the debug output of the adapter and having a look at the bulb?

Thanks,
Florian

**Comment by allofmex** [comment]
> brightness level of 1 as expected, but that value seems to switch my bulbs off instead of keeping it on at a very low level.

I can confirm this behavior. Needs more investigation. What I saw so far is that the bulbs internal level is `1` but state is `OFF` after writing 1 to brightness object. If I write level 1 via DeveloperTab, the bulb does NOT switch off and state stays `ON`. But is seems herdsman converter is correct, only level (and transtime) is published. 
Don't get yet why on-off-state is effected too...

**Comment by allofmex** [comment]
Did some more test:
herdsman uses attribute `moveToLevelWithOnOff`, wich is correct. (my last test was with `moveToLevel`).
As far as I understand https://www.nxp.com/docs/en/user-guide/JN-UG-3115.pdf#G22.1048228

- level 0 is not used in zigbee at all, so far an invalid value, however hue still accepts it as off. 
- level 1 is "minimum possible level for the device", this is somehow `off`. (guess this is interpreted different by other devices..)
- results in level 2 = minimum on/brightness level

It worked before because of the error @Strunzdesign found above (min. written was 3).

@Strunzdesign maybe your fix should be updated to map 1-100 to 2-254 instead 1-254. 

By the way: I dont see a difference between level 2 or 3 on a gu10 hue bulb. Sadly :-(

**Comment by Strunzdesign** [comment]
Tank you very much for having a deeper look into this! Yesterday I tried to dig into the code but didn't find anything that may cause this, but your link is interesting. I'm going to have a look at it today and want to conduct some tests just to be sure about the mechanics involved. Furthermore, I want to do a side-by-side comparison using two identical HUE bulbs: one attached via the HUE bridge and the HUE adapter, the other using this ZigBee adapter. Just to see which range of values a HUE bulb supports and whether I see any "jumps" in brightness if going from one level to an adjacent one.

**Comment by Strunzdesign** [comment]
However, you mentioned that you set a value of "1" using the DeveloperTab, and hereby the bulb did not switch off. Ignoring the fact that "1" may be interpreted differently by bulbs, why does "the adapter" additionally set the bulb state to OFF?

**Comment by Strunzdesign** [comment]
It would be nice to be sure that ALL ZigBee devices interpret a level of 1 as OFF. If we change this logic in a generic way by fixing the equations that my patch introduced, it would be BAD if bug reports arise that either (a) indicate that some bulbs can not be switched off anymore (then 1 is interpreted wrongly as a very low dim value) or (b) some bulbs ignore a dim-to-zero command as they are more strict than HUE bulbs and drop commands with a level value of 0. Currently, the code maps a value of 0 to 0, which is a forbidden value according to the spec.

In short, is it safe to map adapter levels [0...100] to bulb levels [1...254]? And to map bulb levels [1...254] to adapter levels [0...100]. Furthermore, if a bulb delivers "0" or "255", should we throw a warning and/or map that to adpater level "0" and "100", respectively?

If you consider this safe, and given the above mentioned documentation you liked to, the fix should be straight forward (I can do it). The most important part of the fix would be a comment describing this behavior :-)

**Comment by allofmex** [comment]
> you mentioned that you set a value of "1" using the DeveloperTab, and hereby the bulb did not switch off.

This was my fault, I used command `moveToLevel` instead `moveToLevelWithOnOff` - never mind.


> It would be nice to be sure that ALL ZigBee devices interpret a level of 1 as OFF.


As expected, it is interpreted different per manufacturer :-(

Tested with Ikea bulb:

- going down to level 1 is still **ON**.
going up to level 1 switches **ON**
(I don't see a visual difference in brightness between level 1 and 2)

Hue bulb:

- level 1 is **OFF**

My suggest: 
Object state to Zigbee: 0 (of 100) is mapped to 0 (of 254), 1+ (of 100) is mapped to 2+ (of 254). No 1.
Zigbee to object: 0 and 1 (of 254) maps to 0 (of 100). (1 should not occur because it is never written, but just in case)

Should work, else someone would have complained with the old implementation, there we used 0 for off as well and skipped 1 and 2 completely.

**Comment by Strunzdesign** [comment]
Ok, I conducted some tests. For that, I took two HUE LCT012 bulbs, one using the HUE bridge and the other via the ZigBee adapter.

- I did not see a difference between bri "1" or "2" (of 100) on the HUE adapter.
- HUE: "bri" is clipped at a maximum of "254" matching the specs.
- HUE: a "bri" value of "1" switches the bulb on but keeps "level" at "0"! Probably an artifact of the HUE adapter we shall ignore.
- HUE: a "bri" value of "2" calculates a level of "1" (but without a visible difference to bri "1").
- HUE: a "level" value of "1" calculates a bri level value of "3".

Thus, the transformation from bri(2) to level(1) to bri(3) shows that the equations are not symmetric / balanced. Furthermore, no idea if the values do change or not... maybe one could measure it, but I'm not able anymore to see a difference (maybe I used a different type of bulb last time).

Thus, I come to the conclusion to either revert my previous patch (the old behavior was not as bad as the current behavior) or to make it right now and add the following:
- Edit the equation according to your mapping proposal together with some descriptive comments describing the results of your research,
- maybe create two Javascript functions for conversion of adapter-to-bulb levels and vice-versa, and re-use that to align and unify those 4-5 places where it is used currently. This would also ease the task of providing a good comment without duplicating it all over the place.

Then there is still a problems if a bulb has "1" for off and "2" as the minimal dim value, but offers its maximum at an arbitrary value such as "160" instead of "254". Why should one do that? No idea, but the spec allows that IIRC.

Regards,
Florian

**Comment by Strunzdesign** [comment]
Working on it now

**Comment by arteck** [comment]
@Strunzdesign we have with your change problems with a tradfri courtain device..

**Comment by Strunzdesign** [comment]
Hi,

I don't understand... that discussion is about a courtain device that has nothing to to with brightness levels... or does it?

Regards,
Florian

**Comment by Strunzdesign** [comment]
Regarding the discussion it is very unlikely that this patch is responsible for that behavior:
- My patch does not touch any "position" datapoints, only "brightness" and "white_value".
- My patch has a symmetric conversion from bulb- to adapter levels and vice-versa. Given that someone writes "90%" to a (curtain) device and reads "10%" back is a clear indication that the issue is located somewhere else.

Regards,
Florian

**Comment by Strunzdesign** [comment]
BTW, as adapter version 1.2.0 is now available in ioBroker I confirm that it works for me and that this bug can be closed. However, I do not close it now due to the ongoing discussion with that curtain device. Feel free to close it.

**Comment by arteck** [comment]
- My patch does not touch any "position" datapoints, only "brightness" and "white_value".

yep i see .. its only for light...sry
